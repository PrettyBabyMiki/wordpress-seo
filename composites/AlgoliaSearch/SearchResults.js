import React from "react";
import PropTypes from "prop-types";
import a11ySpeak from "a11y-speak";
import styled from "styled-components";
import { ZebrafiedListTable } from "../basic/Table/ListTable";
import { Row } from "../basic/Table/Row";
import colors from "../../style-guide/colors.json";

/**
 * The title of the search result item.
 */
const SearchResultTitle = styled.h3`
	padding-left: 10px;
	margin: 0;
	font-size: 1em;
	font-weight: normal;
`;

/**
 * The link to the search result.
 */
const SearchResultLink = styled.a`
	color: ${ colors.$color_black };
	
	&:hover, &:focus {
		color: ${ colors.$color_pink_dark };
	}
`;

/**
 * Returns a single search result item.
 *
 * @param {Object} props Properties of the SearchResult component.
 * @returns {ReactElement} The rendered search result.
 */
export function SearchResult( props ) {
	let post = props.post;

	return (
		<Row {...props}>
			<SearchResultLink href={ post.permalink } onClick={ props.handler }>
				<div>
					<SearchResultTitle>{ post.post_title }</SearchResultTitle>
				</div>
			</SearchResultLink>
		</Row>
	);
}

SearchResult.propTypes = {
	handler: PropTypes.func.isRequired,
	post: PropTypes.object.isRequired,
};

/**
 * Wraps the search results.
 */
const SearchResultsWrapper = styled.div`
	margin-top: 20px;
	width: 100%;
	display: block;
	clear: both;
`;

/**
 * The no results text paragraph.
 */
const NoResults = styled.p`
	margin-left: 10px;
`;

/**
 * Create the ReactElement to render a list of search results.
 *
 * @param {object} props The React props.
 *
 * @returns {ReactElement} A list of search results.
 *
 * @constructor
 */
class SearchResults extends React.Component {

	constructor( props ) {
		super( props );

		this.state = {
			results: props.results,
		};
	}

	/**
	 * Handles the situation that there are zero results.
	 *
	 * @returns {ReactElement|null} A message stating that no results were found or null if no search term was provided.
	 */
	handleZeroResults() {
		if ( this.props.searchString !== "" ) {
			return this.renderNoResultsFound();
		}

		return null;
	}

	/**
	 * Renders a no results found text.
	 *
	 * @returns {ReactElement} The no results found text.
	 */
	renderNoResultsFound() {
		a11ySpeak( this.props.noResultsText );

		return ( <NoResults>{ this.props.noResultsText }</NoResults> );
	}

	/**
	 * Maps the results to SearchResult components.
	 *
	 * @param {Object} results The results returned by Algolia.
	 *
	 * @returns {Array} Array containing the mapped search results.
	 */
	resultsToSearchItem( results ) {
		return results.map( ( result, index ) => {
			return <SearchResult
				height="24px"
				key={ result.objectID }
				post={ result }
				handler={ ( event ) => {
					event.preventDefault();
					this.props.handler( index );
				} }
			/>;
		} );
	}

	/**
	 * Renders the search results list.
	 *
	 * @returns {ReactElement} A div with either the search results, or a div with a message that no results were found.
	 */
	render() {
		let resultsCount = this.props.results.length;

		// Check whether no results are returned.
		if ( resultsCount <= 0 ) {
			return this.handleZeroResults();
		}

		a11ySpeak( this.props.foundResultsText.replace( "%d", resultsCount ) );

		return (
			<SearchResultsWrapper>
				<ZebrafiedListTable>
					{ this.resultsToSearchItem( this.props.results ) }
				</ZebrafiedListTable>
			</SearchResultsWrapper>
		);
	}

}

SearchResults.propTypes = {
	post: PropTypes.object.isRequired,
	showDetail: PropTypes.func.isRequired,
	handler: PropTypes.func.isRequired,
	searchString: PropTypes.string,
	results: PropTypes.array,
	noResultsText: PropTypes.string.isRequired,
	foundResultsText: PropTypes.string.isRequired,
};

SearchResults.defaultProps = {
	searchString: "",
	results: [],
};

export default SearchResults;
