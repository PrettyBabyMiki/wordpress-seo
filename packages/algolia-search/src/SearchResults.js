/* External dependencies */
import React from "react";
import PropTypes from "prop-types";
import { speak as a11ySpeak } from "@wordpress/a11y";
import styled from "styled-components";
import { __, sprintf } from "@wordpress/i18n";

/* Yoast dependencies */
import { ZebrafiedListTable, Row } from "@yoast/components";
import { colors } from "@yoast/style-guide";

/**
 * The title of the search result item.
 */
const SearchResultTitle = styled.p`
	margin: 0;
	font-size: 1em;
	font-weight: normal;
`;

/**
 * The link to the search result.
 */
const SearchResultLink = styled.a`
	color: ${ colors.$color_black };
	padding: 8px 16px;

	&:hover, &:focus {
		color: ${ colors.$color_pink_dark };
	}
`;

/**
 * Returns a single search result item.
 *
 * @param {Object} props Properties of the SearchResult component.
 * @returns {ReactElement} The rendered search result.
 */
export function SearchResult( props ) {
	const post = props.post;

	return (
		<Row { ...props }>
			<SearchResultLink href={ post.permalink } onClick={ props.onClick }>
				<SearchResultTitle>{ post.post_title }</SearchResultTitle>
			</SearchResultLink>
		</Row>
	);
}

SearchResult.propTypes = {
	onClick: PropTypes.func.isRequired,
	post: PropTypes.object.isRequired,
};

/**
 * Wraps the search results.
 */
const SearchResultsWrapper = styled.div`
	margin-top: 20px;
	clear: both;
`;

/**
 * The no results text paragraph.
 */
const NoResults = styled.p`
	margin-left: 10px;
`;

/**
 * Creates the SearchResults component to render a list of search results.
 *
 * @param {object} props The React props.
 *
 * @returns {ReactElement} The SearchResults component.
 */
class SearchResults extends React.Component {
	/**
	 * Constructs the component and sets its initial state.
	 *
	 * @param {Object} props The props to use for this component.
	 *
	 * @constructor
	 */
	constructor( props ) {
		super( props );

		this.state = {
			results: props.results,
		};
	}

	/**
	 * Handles the situation that there are zero results.
	 *
	 * @returns {ReactElement|null} A message stating that no results were found or null if no search term was provided.
	 */
	handleZeroResults() {
		if ( this.props.searchString !== "" ) {
			return this.renderNoResultsFound();
		}

		return null;
	}

	/**
	 * Renders a no results found text.
	 *
	 * @returns {ReactElement} The no results found text.
	 */
	renderNoResultsFound() {
		const noResultsText = __( "No results found.", "yoast-components" );

		a11ySpeak( noResultsText );

		return (
			<NoResults>
				{ noResultsText }
			</NoResults>
		);
	}

	/**
	 * Maps the results to SearchResult components.
	 *
	 * @param {Object} results The results returned by Algolia.
	 *
	 * @returns {Array} Array containing the mapped search results.
	 */
	resultsToSearchItem( results ) {
		return results.map( ( result, index ) => {
			return <SearchResult
				rowHeight="32px"
				key={ result.objectID }
				post={ result }
				// Note: this passes the onClick but actually also attaches a click event on the LI element.
				onClick={ ( event ) => {
					event.preventDefault();
					event.stopPropagation();
					this.props.onClick( index );
				} }
			/>;
		} );
	}

	/**
	 * Renders the search results list.
	 *
	 * @returns {ReactElement} Either returns the search results, or a message that no results were found.
	 */
	render() {
		const resultsCount = this.props.results.length;

		// Check whether no results are returned.
		if ( resultsCount <= 0 ) {
			return this.handleZeroResults();
		}

		const foundResultsText = sprintf( __( "Number of results found: %d", "yoast-components" ), resultsCount );

		a11ySpeak( foundResultsText );

		return (
			<SearchResultsWrapper>
				<h2 className="screen-reader-text">
					{ __( "Search results", "yoast-components" ) }
				</h2>
				<ZebrafiedListTable>
					{ this.resultsToSearchItem( this.props.results ) }
				</ZebrafiedListTable>
			</SearchResultsWrapper>
		);
	}
}

SearchResults.propTypes = {
	onClick: PropTypes.func.isRequired,
	searchString: PropTypes.string,
	results: PropTypes.array,
};

SearchResults.defaultProps = {
	searchString: "",
	results: [],
};

export default SearchResults;
