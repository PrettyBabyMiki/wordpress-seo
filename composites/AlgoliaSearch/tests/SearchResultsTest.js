import React from "react";
import renderer from "react-test-renderer";

import SearchResults from "../SearchResults.js";

let post = { permalink: "https://kb.yoast.com/kb/passive-voice/", postTitle: "Post Title", objectID: 1 };

test( "the SearchResults component with results matches the snapshot", () => {
	const component = renderer.create(
		<SearchResults handler={ () => {} }
					   post={ post }
					   showDetail={ () => {} }
					   searchString="Test"
					   results={ [ post ] }
					   noResultsText="No Results"
					   foundResultsText="These are your %d results"
		/>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the SearchResults component without results matches the snapshot", () => {
	const component = renderer.create(
		<SearchResults handler={ () => {} }
					   post={ {} }
					   showDetail={ () => {} }
					   searchString="Test"
					   results={ [] }
					   noResultsText="No Results"
					   foundResultsText="These are your %d results"
		/>
	);

	let tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
