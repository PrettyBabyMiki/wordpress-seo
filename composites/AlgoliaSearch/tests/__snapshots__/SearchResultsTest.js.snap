// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the SearchResults component with results matches the snapshot 1`] = `
.c1 {
  margin: 0;
  padding: 0;
  list-style: none;
  position: relative;
  width: 100%;
}

.c1 li:first-child > span::before {
  left: auto;
}

.c2 {
  background: #f7f7f7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 32px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  margin: 0;
  font-size: 1em;
  font-weight: normal;
}

.c3 {
  color: #000;
  padding: 8px 16px;
}

.c3:hover,
.c3:focus {
  color: #a4286a;
}

.c0 {
  margin-top: 20px;
  clear: both;
}

<div
  className="c0"
>
  <h2
    className="screen-reader-text"
  >
    Search results
  </h2>
  <ul
    className="c1"
    role="list"
  >
    <li
      className="c2"
      onClick={[Function]}
    >
      <a
        className="c3"
        href="https://kb.yoast.com/kb/passive-voice/"
        onClick={[Function]}
      >
        <p
          className="c4"
        />
      </a>
    </li>
  </ul>
</div>
`;

exports[`the SearchResults component without results matches the snapshot 1`] = `
.c0 {
  margin-left: 10px;
}

<p
  className="c0"
>
  No results found.
</p>
`;
